File: implementations/typescript/src/parser.ts
import { AstNode, BlockNode, DocumentNode } from "./ast.js";
import { ParserException } from "./error.js";
import { Lexer } from "./lexer.js";
import { Token, TokenType } from "./token.js";

export class Parser {
  private current: Token;

  constructor(private readonly lexer: Lexer) {
    this.current = lexer.nextToken();
  }

  parse(): DocumentNode {
    if (this.current.type === TokenType.Do) {
      this.eat(TokenType.Do);
      const rootValue = this.parseArray();
      if (this.current.type !== TokenType.EOF) throw this.err("unexpected token after root array");
      return { kind: "document", blocks: [], root_value: rootValue };
    }

    const blocks: BlockNode[] = [];
    while (this.current.type !== TokenType.EOF) blocks.push(this.parseBlock());
    return { kind: "document", blocks };
  }

  private parseBlock(): BlockNode {
    const name = this.current.value;
    this.eat(TokenType.Identifier);

    let argument: string | undefined;
    if (this.current.type === TokenType.String) {
      argument = this.current.value;
      this.eat(TokenType.String);
    }

    this.eat(TokenType.Do);

    const { properties, blocks, namedBlocks } = this.parseBlockBody();

    this.eat(TokenType.End);
    const block: BlockNode = { kind: "block", name, properties, blocks };
    if (argument !== undefined) block.argument = argument;
    if (namedBlocks.length > 0) block.named_blocks = namedBlocks;
    return block;
  }

  private parseAnonymousBlock(): BlockNode {
    this.eat(TokenType.Do);
    const { properties, blocks, namedBlocks } = this.parseBlockBody();
    this.eat(TokenType.End);
    const block: BlockNode = { kind: "block", name: "", properties, blocks };
    if (namedBlocks.length > 0) block.named_blocks = namedBlocks;
    return block;
  }

  private parseBlockBody(): { properties: Record<string, AstNode>; blocks: Record<string, BlockNode>; namedBlocks: BlockNode[] } {
    const properties: Record<string, AstNode> = {};
    const blocks: Record<string, BlockNode> = {};
    const namedBlocks: BlockNode[] = [];
    const seenKeys = new Set<string>();

    while (this.current.type !== TokenType.End) {
      if (this.current.type === TokenType.EOF) throw this.err("missing 'end' for block");
      if (this.current.type !== TokenType.Identifier) throw this.err(`expected identifier, got ${this.current.type}`);

      const next = this.peekToken();
      if (next.type === TokenType.Do) {
        const afterDo = this.peekToken(2);
        if (afterDo.type === TokenType.LBracket) {
          const key = this.current.value;
          this.eat(TokenType.Identifier);
          this.ensurePropertyKeyValid(key, seenKeys);
          this.eat(TokenType.Do);
          properties[key] = this.parseArray();
          this.eat(TokenType.End);
        } else {
          const child = this.parseBlock();
          if (child.argument) namedBlocks.push(child);
          else blocks[child.name] = child;
        }
      } else if (next.type === TokenType.String) {
        const child = this.parseBlock();
        if (child.argument) namedBlocks.push(child);
        else blocks[child.name] = child;
      } else if (next.type === TokenType.Equal || next.type === TokenType.Dot) {
        const key = this.parsePropertyKey();
        this.ensurePropertyKeyValid(key, seenKeys);
        this.eat(TokenType.Equal);
        properties[key] = this.parseValue();
      } else throw this.err(`invalid statement after '${this.current.value}'`);
    }

    return { properties, blocks, namedBlocks };
  }

  private parsePropertyKey(): string {
    let key = this.current.value;
    this.eat(TokenType.Identifier);
    for (;;) {
      const curType = this.current.type;
      if (curType !== TokenType.Dot) break;
      this.eat(TokenType.Dot);
      const afterDot = this.current.type;
      if (afterDot !== TokenType.Identifier) throw this.err("expected identifier after dot");
      key += `.${this.current.value}`;
      this.eat(TokenType.Identifier);
    }
    return key;
  }

  private parseValue(): AstNode {
    if (this.current.type === TokenType.String) {
      const v = this.current.value; this.eat(TokenType.String); return { kind: "string", value: v };
    }
    if (this.current.type === TokenType.Number) {
      const v = Number(this.current.value); this.eat(TokenType.Number); return { kind: "number", value: v };
    }
    if (this.current.type === TokenType.Identifier) {
      const v = this.current.value; this.eat(TokenType.Identifier);
      if (v === "true") return { kind: "boolean", value: true };
      if (v === "false") return { kind: "boolean", value: false };
      throw this.err(`invalid bare value '${v}'`);
    }
    if (this.current.type === TokenType.LBracket) return this.parseArray();
    if (this.current.type === TokenType.Do) return this.parseAnonymousBlock();
    throw this.err(`unexpected token ${this.current.type}`);
  }

  private parseArray(): AstNode {
    this.eat(TokenType.LBracket);
    const elements: AstNode[] = [];
    if (this.current.type !== TokenType.RBracket) {
      elements.push(this.parseValue());
      for (;;) {
        const curType = this.current.type;
        if (curType !== TokenType.Comma) break;
        const nextType = this.peekToken().type;
        if (nextType === TokenType.RBracket) throw this.err("trailing comma in array");
        this.eat(TokenType.Comma);
        elements.push(this.parseValue());
      }
    }
    this.eat(TokenType.RBracket);
    return { kind: "array", elements };
  }

  private eat(type: TokenType): void {
    if (this.current.type !== type) throw this.err(`expected ${type}, got ${this.current.type}`);
    this.current = this.lexer.nextToken();
  }

  private peekToken(offset = 1): Token {
    const st = this.lexer.snapshot();
    let tok = this.current;
    for (let i = 0; i < offset; i += 1) tok = this.lexer.nextToken();
    this.lexer.restore(st);
    return tok;
  }

  private err(message: string): ParserException {
    return new ParserException(message, this.current.line, this.current.column);
  }

  private ensurePropertyKeyValid(key: string, seen: Set<string>): void {
    if (seen.has(key)) throw this.err(`duplicate key '${key}'`);
    const parts = key.split(".");
    for (const existing of seen) {
      const ex = existing.split(".");
      if (this.isPrefix(parts, ex) || this.isPrefix(ex, parts)) throw this.err(`key conflict '${key}' vs '${existing}'`);
    }
    seen.add(key);
  }

  private isPrefix(left: string[], right: string[]): boolean {
    if (left.length >= right.length) return false;
    for (let i = 0; i < left.length; i += 1) if (left[i] !== right[i]) return false;
    return true;
  }
}

File: implementations/go/parser.go
package rcl

import (
	"strconv"
	"strings"
)

type parser struct {
	lex *lexer
	cur token
}

func Parse(text string) (DocumentNode, error) {
	p := &parser{lex: newLexer(text)}
	t, err := p.lex.nextToken()
	if err != nil { return DocumentNode{}, err }
	p.cur = t
	return p.parse()
}

func (p *parser) parse() (DocumentNode, error) {
	if p.cur.typ == tokDo {
		if err := p.eat(tokDo); err != nil { return DocumentNode{}, err }
		root, err := p.parseArray()
		if err != nil { return DocumentNode{}, err }
		if p.cur.typ != tokEOF { return DocumentNode{}, p.err("unexpected token after root array") }
		return DocumentNode{NodeKind: "document", Blocks: []BlockNode{}, RootValue: root}, nil
	}

	blocks := []BlockNode{}
	for p.cur.typ != tokEOF {
		b, err := p.parseBlock()
		if err != nil { return DocumentNode{}, err }
		blocks = append(blocks, b)
	}
	return DocumentNode{NodeKind: "document", Blocks: blocks}, nil
}

func (p *parser) parseBlock() (BlockNode, error) {
	name := p.cur.value
	if err := p.eat(tokIdentifier); err != nil { return BlockNode{}, err }
	var arg *string
	if p.cur.typ == tokString { v := p.cur.value; arg = &v; _ = p.eat(tokString) }
	if err := p.eat(tokDo); err != nil { return BlockNode{}, err }

	props, blocks, named, err := p.parseBlockBody()
	if err != nil { return BlockNode{}, err }
	if err := p.eat(tokEnd); err != nil { return BlockNode{}, err }
	b := BlockNode{NodeKind: "block", Name: name, Argument: arg, Properties: props, Blocks: blocks}
	if len(named) > 0 { b.NamedBlocks = named }
	return b, nil
}

func (p *parser) parseAnonymousBlock() (BlockNode, error) {
	if err := p.eat(tokDo); err != nil { return BlockNode{}, err }
	props, blocks, named, err := p.parseBlockBody()
	if err != nil { return BlockNode{}, err }
	if err := p.eat(tokEnd); err != nil { return BlockNode{}, err }
	b := BlockNode{NodeKind: "block", Name: "", Properties: props, Blocks: blocks}
	if len(named) > 0 { b.NamedBlocks = named }
	return b, nil
}

func (p *parser) parseBlockBody() (map[string]Node, map[string]BlockNode, []BlockNode, error) {
	props := map[string]Node{}
	blocks := map[string]BlockNode{}
	named := []BlockNode{}
	seen := map[string]bool{}

	for p.cur.typ != tokEnd {
		if p.cur.typ == tokEOF { return nil, nil, nil, p.err("missing end") }
		if p.cur.typ != tokIdentifier { return nil, nil, nil, p.err("expected identifier") }

		next, err := p.peekAt(1)
		if err != nil { return nil, nil, nil, err }

		if next.typ == tokDo {
			afterDo, err := p.peekAt(2)
			if err != nil { return nil, nil, nil, err }
			if afterDo.typ == tokLBracket {
				key := p.cur.value
				if err := p.eat(tokIdentifier); err != nil { return nil, nil, nil, err }
				if err := p.ensureKeyValid(key, seen); err != nil { return nil, nil, nil, err }
				if err := p.eat(tokDo); err != nil { return nil, nil, nil, err }
				v, err := p.parseArray()
				if err != nil { return nil, nil, nil, err }
				props[key] = v
				if err := p.eat(tokEnd); err != nil { return nil, nil, nil, err }
			} else {
				child, err := p.parseBlock()
				if err != nil { return nil, nil, nil, err }
				if child.Argument != nil { named = append(named, child) } else { blocks[child.Name] = child }
			}
		} else if next.typ == tokString {
			child, err := p.parseBlock()
			if err != nil { return nil, nil, nil, err }
			if child.Argument != nil { named = append(named, child) } else { blocks[child.Name] = child }
		} else if next.typ == tokEqual || next.typ == tokDot {
			k, err := p.parseKey()
			if err != nil { return nil, nil, nil, err }
			if err = p.ensureKeyValid(k, seen); err != nil { return nil, nil, nil, err }
			if err = p.eat(tokEqual); err != nil { return nil, nil, nil, err }
			v, err := p.parseValue()
			if err != nil { return nil, nil, nil, err }
			props[k] = v
		} else { return nil, nil, nil, p.err("invalid statement") }
	}

	return props, blocks, named, nil
}

func (p *parser) parseKey() (string, error) {
	key := p.cur.value
	if err := p.eat(tokIdentifier); err != nil { return "", err }
	for p.cur.typ == tokDot {
		if err := p.eat(tokDot); err != nil { return "", err }
		if p.cur.typ != tokIdentifier { return "", p.err("expected identifier after dot") }
		key += "." + p.cur.value
		if err := p.eat(tokIdentifier); err != nil { return "", err }
	}
	return key, nil
}

func (p *parser) parseValue() (Node, error) {
	switch p.cur.typ {
	case tokString:
		v := StringNode{NodeKind: "string", Value: p.cur.value}; _ = p.eat(tokString); return v, nil
	case tokNumber:
		n, err := strconv.ParseFloat(p.cur.value, 64)
		if err != nil { return nil, p.err("invalid number") }
		_ = p.eat(tokNumber); return NumberNode{NodeKind: "number", Value: n}, nil
	case tokIdentifier:
		v := p.cur.value; _ = p.eat(tokIdentifier)
		if v == "true" { return BooleanNode{NodeKind: "boolean", Value: true}, nil }
		if v == "false" { return BooleanNode{NodeKind: "boolean", Value: false}, nil }
		return nil, p.err("invalid bare value")
	case tokLBracket:
		return p.parseArray()
	case tokDo:
		b, err := p.parseAnonymousBlock()
		if err != nil { return nil, err }
		return b, nil
	default:
		return nil, p.err("unexpected value")
	}
}

func (p *parser) parseArray() (Node, error) {
	if err := p.eat(tokLBracket); err != nil { return nil, err }
	elems := []Node{}
	if p.cur.typ != tokRBracket {
		v, err := p.parseValue(); if err != nil { return nil, err }
		elems = append(elems, v)
		for p.cur.typ == tokComma {
			_ = p.eat(tokComma)
			if p.cur.typ == tokRBracket { return nil, p.err("trailing comma in array") }
			v, err = p.parseValue(); if err != nil { return nil, err }
			elems = append(elems, v)
		}
	}
	if err := p.eat(tokRBracket); err != nil { return nil, err }
	return ArrayNode{NodeKind: "array", Elements: elems}, nil
}

func (p *parser) eat(tt tokenType) error {
	if p.cur.typ != tt { return p.err("unexpected token") }
	n, err := p.lex.nextToken()
	if err != nil { return err }
	p.cur = n
	return nil
}
func (p *parser) peekAt(offset int) (token, error) {
	st := p.lex.snapshot()
	t := p.cur
	var err error
	for i := 0; i < offset; i++ {
		t, err = p.lex.nextToken()
		if err != nil {
			p.lex.restore(st)
			return token{}, err
		}
	}
	p.lex.restore(st)
	return t, nil
}
func (p *parser) err(msg string) error { return &ParseError{Message: msg, Line: p.cur.line, Column: p.cur.col} }

func (p *parser) ensureKeyValid(key string, seen map[string]bool) error {
	if seen[key] { return p.err("duplicate key") }
	parts := strings.Split(key, ".")
	for existing := range seen {
		ex := strings.Split(existing, ".")
		if isPrefix(parts, ex) || isPrefix(ex, parts) { return p.err("key conflict") }
	}
	seen[key] = true
	return nil
}

func isPrefix(left, right []string) bool {
	if len(left) >= len(right) { return false }
	for i := range left { if left[i] != right[i] { return false } }
	return true
}
